crates/sound/src/device.rs

```diff
@@ -70,49 +76,301 @@ impl VhostUserSoundThread {
+                            },
+                        )
+                        .unwrap();
+
+                        // PcmRelease needs to flush IO messages; the audio backend will reply when
+                        // it drops the ControlMessage.
+                        continue;
+                    }
+                }
+                ControlMessageKind::PcmStart => {
+                    let request = desc_chain
+                        .memory()
+                        .read_obj::<VirtioSoundPcmHeader>(desc_request.addr())
+                        .map_err(|_| Error::DescriptorReadFailed)?;
+                    let stream_id = request.stream_id.into();
+
+                    if stream_id as usize >= self.streams_no {
+                        log::error!("{}", Error::from(StreamError::InvalidStreamId(stream_id)));
+                        resp.code = VIRTIO_SND_S_BAD_MSG.into();
+                    } else {
+                        let b = audio_backend.write().unwrap();
+                        b.start(stream_id).unwrap();
+                    }
+                }
+                ControlMessageKind::PcmStop => {
+                    let request = desc_chain
+                        .memory()
+                        .read_obj::<VirtioSoundPcmHeader>(desc_request.addr())
+                        .map_err(|_| Error::DescriptorReadFailed)?;
+                    let stream_id = request.stream_id.into();
+
+                    if stream_id as usize >= self.streams_no {
+                        log::error!("{}", Error::from(StreamError::InvalidStreamId(stream_id)));
+                        resp.code = VIRTIO_SND_S_BAD_MSG.into();
+                    } else {
+                        let b = audio_backend.write().unwrap();
+                        b.stop(stream_id).unwrap();
+                    }
+                }
+                ControlMessageKind::ChmapInfo => {
+                    resp.code = VIRTIO_SND_S_NOT_SUPP.into();
+                }
+            }
+            log::trace!(
+                "returned {} for ctrl msg {:?}",
+                match u32::from(resp.code) {
+                    v if v == VIRTIO_SND_S_OK => "OK",
+                    v if v == VIRTIO_SND_S_BAD_MSG => "BAD_MSG",
+                    v if v == VIRTIO_SND_S_NOT_SUPP => "NOT_SUPP",
+                    v if v == VIRTIO_SND_S_IO_ERR => "IO_ERR",
+                    _ => unreachable!(),
+                },
+                code
+            );
+            desc_chain
+                .memory()
+                .write_obj(resp, desc_hdr.addr())
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+
+            if vring.add_used(desc_chain.head_index(), len).is_err() {
+                log::error!("Couldn't return used descriptors to the ring");
+            }
+            any |= true;
+        }
+
+        // Send notification if any request was processed
+        if any && vring.signal_used_queue().is_err() {
```

Comment:
In which cases can happen that we get here and `any` is false?

Reply:
If the request buffer is empty, there should be an empty descriptor or lack of descriptors?, which the spec doesn't explicitly state (Again, I think, I am not exactly sure) the relevant part is this:

> An I/O message consists of the header part, followed by the buffer, and then the status part. 

I'd expect there would be an empty descriptor instead of none but I couldn't find any such guarantee in the code, or the spec

Reply:
this might be clarified somewhere else that specifies the mapping of payload to virt queue descriptors.

Reply:
If I understand correctly, any is `false` if the code does not go into the for-loop (https://github.com/epilys/vhost-device/blob/95ca8a85f12bc52946e022c9d94cf89d75d40666/crates/sound/src/device.rs#L143). However, I wonder if the codes does not exit before at https://github.com/epilys/vhost-device/blob/95ca8a85f12bc52946e022c9d94cf89d75d40666/crates/sound/src/device.rs#L132.